src/main/java/io/axoniq/axonserver/connector/impl/FlowControlledBuffer.java

```diff
@@ -111,11 +118,36 @@ protected T tryTakeNow() {
      *
      * @param timeout  the duration to wait for an entry to become available in the buffer
      * @param timeUnit the {@link TimeUnit} used to specify the duration together with the {@code timeout}
-     * @return an entry of type {@code T} from this buffer if present, otherwise {@code null}
+     * @return an entry of type {@code T} from this buffer if present, otherwise {@code null}. Timeouts will result in
+     * null as well
      * @throws InterruptedException while waiting for an entry to be taken
      */
     protected T tryTake(long timeout, TimeUnit timeUnit) throws InterruptedException {
-        T taken = validate(buffer.poll(timeout, timeUnit), true);
+        try {
+            return tryTake(timeout, timeUnit, false);
+        } catch (TimeoutException e) {
+            return null;
+        }
+    }
+
+    /**
+     * Try to retrieve an entry of type {@code T} from the buffer, waiting for the duration of {@code timeout} in the
+     * given {@code timeUnit}. If none is present, {@code null} will be returned.
+     *
+     * @param timeout            the duration to wait for an entry to become available in the buffer
+     * @param timeUnit           the {@link TimeUnit} used to specify the duration together with the {@code timeout}
+     * @param exceptionOnTimeout Whether a {@code TimeoutException} should be thrown when the operation times out
+     * @return an entry of type {@code T} from this buffer if present, otherwise {@code null}
+     * @throws InterruptedException while waiting for an entry to be taken
+     * @throws TimeoutException     If there is no message available after waiting the allotted period
+     */
+    protected T tryTake(long timeout, TimeUnit timeUnit, boolean exceptionOnTimeout)
```

Comment:
This API forces a component to catch `TimeoutException` even when `exceptionOnTimeout` is false.

Reply:
I changed the API, having a dedicated RuntimeException similar for other situations we already have one for.